### DISCUSSION TITLE
Can duckdb read csv bytes array in memory ?

### DISCUSSION THREAD

**Discussion by LiangJiechao** [duckdb/duckdb#8985] [Q&A] [answered]
Can duckdb read csv bytes array in memory ? If can, please tell me how to do, thank.
Because the server return the csv bytes array, now I save them to the disk, and then duckdb read the csv files to insert the data.

**Comment by Tishj** [comment, marked as answer]
So far only the Python client has this ability I believe.
It uses the `fsspec` package to store the data for the duration of the query.
```py
import duckdb

con = duckdb.connect()
string = StringIO("c1,c2,c3\na,b,c")
res = con.read_csv(string, header=True).fetchall()
assert res == [('a', 'b', 'c')]
```

**Reply by LiangJiechao** [reply]
Thanks for your reply, but my project's client is C#.
Anyway, thank you.

**Reply by FrancoisLepoutre** [reply]
Hi Tishj

I am using duckdb from python. And yes that's a joy to access duckdn from python.  

res = con.read_csv(string, header=True).fetchall()

Thanks for the tip! I was not aware of this in-memory access from python. And I used pyarrow as a bridge as it offers its own set of feeding schemes. 

The combo proves both fast and solid. It offers a large and flexible array of secure importing schemes including distant and in-memory ones. Even more so if you combine with fsspec . But I value straight access wonder. Thus I wonder whether this straight-from-python in-memory access extend to parquet resources. 

Can you?
res = con.read_csv(myParquetIOString, header=True).fetchall()

**Reply by Tishj** [reply]
The functionality has not been extended to the `read_parquet` method yet, though this should be trivial to do

**Reply by FrancoisLepoutre** [reply]
Thanks a lot for the feedback!
